r/R/dplyr-functions.R

```diff
@@ -280,6 +284,68 @@ nse_funcs$str_trim <- function(string, side = c("both", "left", "right")) {
   Expression$create(trim_fun, string)
 }
 
+nse_funcs$substr <- function(string, start, stop) {
+  assert_that(
+    length(start) == 1,
+    msg = "`start` must be length 1 - other lengths are not supported in Arrow"
+  )
+  assert_that(
+    length(stop) == 1,
+    msg = "`stop` must be length 1 - other lengths are not supported in Arrow"
+  )
+
+  # substr treats values as if they're on a continous number line, so values 
+  # 0 are effectively blank characters - set `start` to 1 here so Arrow mimics 
+  # this behavior
+  if (start <= 0) {
+    start <- 1
+  }
+
+  # if `stop` is lower than `start`, this is invalid, so set `stop` to
+  # 0 so that an empty string will be returned
+  if (stop < start) {
+    stop <- 0
+  }
+
+  Expression$create(
+    "utf8_slice_codeunits",
+    string,
+    options = list(start = start - 1L, stop = stop)
+  )
+}
+
+nse_funcs$substring <- nse_funcs$substr
+
+nse_funcs$str_sub <- function(string, start = 1L, end = -1L) {
+  assert_that(
+    length(start) == 1,
+    msg = "`start` must be length 1 - other lengths are not supported in Arrow"
+  )
+  assert_that(
+    length(end) == 1,
+    msg = "`end` must be length 1 - other lengths are not supported in Arrow"
+  )
+
+  if (end == -1) {
+    end <- .Machine$integer.max
+  }
+
+  if (end < start) {
+    end <- 0
+  }
+
+  
+  if (start > 0) {
+    start <- start - 1L
```

Comment:
These deserve explanatory comments, particularly noting the differences in behavior among base::substr, stringr::str_sub, and arrow's utf8_slice_codeunits. Basically a concise version of what you answered me in the PR review. (It's a good rule of thumb that if I ask you why something is a certain way because it wasn't obvious to me, the answer should probably be a code comment and not (just) a PR comment--if it's not obvious to the reader now, it definitely won't be obvious to us in 6 months or whenever we have to revise this code and wonder why things are munged a certain way.)

Reply:
Thanks, good to know, will update with this